Lumo class is now loaded by default when no Theme is found

gilberto-torrezan opened the pull request:
Restructured the code to place all Theme discovery code in a single place - RouteRegistry.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/flow/3785)
<!-- Reviewable:end -->

denis-anisimov commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kNT8gAMgjbwk4YA8p:-L8kNT8gAMgjbwk4YA8q:b-a2jxmb) (no related file):*
I think we have been discussing the maven plugin behavior as well in regard to this feature.
Bundling should also use Lumo if there is no theme defined.

So please either create a ticket for maven plugin or it should be in this PR.

---

*[flow-server/src/main/java/com/vaadin/flow/component/UI.java, line 691 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kJnZS7XqojPgLFQ5l:-L8kJnZS7XqojPgLFQ5m:b85ycj1) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/component/UI.java#L691)):*
> return registry.getThemeFor(navigationTarget);

There are two calls that do the same thing.
The only difference is : getting the registry.

I would recommend to change the logic so that it sets some `registry` variable and in the end if it's not `null` call  `registry.getThemeFor(navigationTarget);` .

I'm not sure by the way that router is the correct place to store info about theming. But that's another thing.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 73 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kKx2wDC1mpcdk1f2v:-L8kKx2wDC1mpcdk1f2w:bexet3m) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L73)):*
> private static final Class<? extends AbstractTheme> LUMO_CLASS_IF_AVAILABLE;
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>
>     static {
>         Class<? extends AbstractTheme> theme = null;
>         try {
>             theme = (Class<? extends AbstractTheme>) Class
>                     .forName("com.vaadin.flow.theme.lumo.Lumo");
>         } catch (ClassNotFoundException e) {
>             // ignore, the Lumo class is not available in the classpath
>         }
>         LUMO_CLASS_IF_AVAILABLE = theme;
>     }
</details>


Not a serious, but I personally don't like this pattern with a `static` block.
Instead it may be done via `private static` method which does the same as the static block here but returns a value which is used immediately in the static field initialization.
The code becomes clearer.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8kMSlZCOEqKM6px9mJ:bwsscdv) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
> private Class<? extends AbstractTheme> findThemeForNavigationTarget(
>             Class<?> navigationTarget) {
>  
>         Optional<Theme> themeAnnotation = AnnotationReader
>                 .getAnnotationFor(navigationTarget, Theme.class);
>  
>         if (themeAnnotation.isPresent()) {
>             return themeAnnotation.get().value();
>         }
>  
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>
>         if (!AnnotationReader.getAnnotationFor(navigationTarget, NoTheme.class)
>                 .isPresent()) {
>             return LUMO_CLASS_IF_AVAILABLE;
>         }
</details>
>  
>         return null;
>     }

Is this really needed to be pre-calculated ?
Why not just try to get the theme for the target ( where it's requested) and if there is no theme then return `LUMO_CLASS_IF_AVAILABLE`.

By the way I'm not sure that this code is correct. I think theme should be discovered traversing parent layouts. It doesn't have to be necessary defined for the component itself.
@caalador could you please check this also?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

*[flow-server/src/main/java/com/vaadin/flow/component/UI.java, line 691 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kJnZS7XqojPgLFQ5l:-L8kPN8fBMtftbAbtrcz:bn0hq3z) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/component/UI.java#L691)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> return registry.getThemeFor(navigationTarget);

There are two calls that do the same thing.
The only difference is : getting the registry.

I would recommend to change the logic so that it sets some `registry` variable and in the end if it's not `null` call  `registry.getThemeFor(navigationTarget);` .

I'm not sure by the way that router is the correct place to store info about theming. But that's another thing.
</blockquote></details>

Also one point is that if we for some reason do not have a "new" Router then we don't handle theme.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 70 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kPpAP2rcSZ-7HPoO5:-L8kPpAP2rcSZ-7HPoO6:bhudzy7) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L70)):*
> ```Java
>             theme = (Class<? extends AbstractTheme>) Class
>                     .forName("com.vaadin.flow.theme.lumo.Lumo");
>         } catch (ClassNotFoundException e) {
> ```

logg as trace.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8kPtnS9feXx9MP3gyv:bbuxssy) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private Class<? extends AbstractTheme> findThemeForNavigationTarget(
>             Class<?> navigationTarget) {
>  
>         Optional<Theme> themeAnnotation = AnnotationReader
>                 .getAnnotationFor(navigationTarget, Theme.class);
>  
>         if (themeAnnotation.isPresent()) {
>             return themeAnnotation.get().value();
>         }
>  
>         if (!AnnotationReader.getAnnotationFor(navigationTarget, NoTheme.class)
>                 .isPresent()) {
>             return LUMO_CLASS_IF_AVAILABLE;
>         }
>  
>         return null;
>     }

Is this really needed to be pre-calculated ?
Why not just try to get the theme for the target ( where it's requested) and if there is no theme then return `LUMO_CLASS_IF_AVAILABLE`.

By the way I'm not sure that this code is correct. I think theme should be discovered traversing parent layouts. It doesn't have to be necessary defined for the component itself.
@caalador could you please check this also?
</blockquote></details>

Yes the Theme annotation should always be on the top parent. so the correct check is the one done in UIInternals update theme which should be moved here.

The cached theme for a class should not be for the parent target, but for the actual navigation target which should check the topParent first and then if no topParent exists check itself for a `@Theme` or `@NoTheme`

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8kYMPj3SnYEBPtmYkF:b-l1oofy) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Yes the Theme annotation should always be on the top parent. so the correct check is the one done in UIInternals update theme which should be moved here.

The cached theme for a class should not be for the parent target, but for the actual navigation target which should check the topParent first and then if no topParent exists check itself for a `@Theme` or `@NoTheme`
</blockquote></details>

Since some calculation should be done anyway: I would suggest do not cache `LUMO_CLASS_IF_AVAILABLE` at all.
Just keep the previous logic. And then in the end if no theme is calculated : just use `LUMO_CLASS_IF_AVAILABLE`.

The caching logic from here increases the code and use extrac atomic reference with a map inside it which makes everything too cumbersome.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8ke4UE8HKrWnslwHnS:b-6rql2j) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

Since some calculation should be done anyway: I would suggest do not cache `LUMO_CLASS_IF_AVAILABLE` at all.
Just keep the previous logic. And then in the end if no theme is calculated : just use `LUMO_CLASS_IF_AVAILABLE`.

The caching logic from here increases the code and use extrac atomic reference with a map inside it which makes everything too cumbersome.
</blockquote></details>

The info could be stored in the RouteTarget. The main point is that this information will never change after servlet startup and not always searching for it will lessen the load a little.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8kf-kGA7oYhi4NV6VP:bosrz4h) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

The info could be stored in the RouteTarget. The main point is that this information will never change after servlet startup and not always searching for it will lessen the load a little.
</blockquote></details>

`RouteTarget` is OK.
I would like to get rid of additional `AtomicReference`  and a map inside it.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kNT8gAMgjbwk4YA8p:-L8kO8vaCSsrtTDUrKoX:b-1ij7v3) (no related file):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

I think we have been discussing the maven plugin behavior as well in regard to this feature.
Bundling should also use Lumo if there is no theme defined.

So please either create a ticket for maven plugin or it should be in this PR.
</blockquote></details>

The plan was to create a separate PR for the plugin, so this one could be reviewed in the meantime.
But your call.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 70 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kPpAP2rcSZ-7HPoO5:-L8kV3eg8UUWkjA6bwCp:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L70)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

logg as trace.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 73 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kKx2wDC1mpcdk1f2v:-L8kV44d2DV3MMVBdJCz:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L73)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private static final Class<? extends AbstractTheme> LUMO_CLASS_IF_AVAILABLE;
>     static {
>         Class<? extends AbstractTheme> theme = null;
>         try {
>             theme = (Class<? extends AbstractTheme>) Class
>                     .forName("com.vaadin.flow.theme.lumo.Lumo");
>         } catch (ClassNotFoundException e) {
>             // ignore, the Lumo class is not available in the classpath
>         }
>         LUMO_CLASS_IF_AVAILABLE = theme;
>     }


Not a serious, but I personally don't like this pattern with a `static` block.
Instead it may be done via `private static` method which does the same as the static block here but returns a value which is used immediately in the static field initialization.
The code becomes clearer.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8kTGRCBsoAyfsLGLqk:b-vlfz7u) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

`RouteTarget` is OK.
I would like to get rid of additional `AtomicReference`  and a map inside it.
</blockquote></details>

The LUMO_CLASS_IF_AVAILABLE thing is not related to the AtomicReference Map at all.
LUMO_CLASS_IF_AVAILABLE field is just to avoid using `Class.forName` every time it is needed (which is an expensive operation). Since the classpath is not supposed to change, it makes sense to load it only once and forget about it.

The cache used in the AtomicReference map is another thing. It's an optimization to avoid transversing the class hierarchy every time looking for the `@Theme` annotation. Since it's just a cache, the application can live without it. It's your call whether we should keep it or not, I have no strong opinions about it.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 4 files at r2.
Review status: 2 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kNT8gAMgjbwk4YA8p:-L8kkoiD9nNvDRRI735z:b79rds1) (no related file):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

The plan was to create a separate PR for the plugin, so this one could be reviewed in the meantime.
But your call.
</blockquote></details>

OK. Just don't forget it.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8klAMH6xvbYiIv1Z_1:b-m8yrg5) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

The LUMO_CLASS_IF_AVAILABLE thing is not related to the AtomicReference Map at all.
LUMO_CLASS_IF_AVAILABLE field is just to avoid using `Class.forName` every time it is needed (which is an expensive operation). Since the classpath is not supposed to change, it makes sense to load it only once and forget about it.

The cache used in the AtomicReference map is another thing. It's an optimization to avoid transversing the class hierarchy every time looking for the `@Theme` annotation. Since it's just a cache, the application can live without it. It's your call whether we should keep it or not, I have no strong opinions about it.
</blockquote></details>

If it's not a cache then may be you can make it a variable which is transfer as a context instead of keeping it as a field ?

If it's not possible then I would just don't use the cache at all.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 2 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8kpDha4XwrFuOlSmaF:b2xmfmw) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

If it's not a cache then may be you can make it a variable which is transfer as a context instead of keeping it as a field ?

If it's not possible then I would just don't use the cache at all.
</blockquote></details>

What? I'm confused. Are you talking about the map? If so, yes it IS a cache. We compute the targetRoutes on init and save the paths in the cache, so the same thing is done with the themes. We can transverse the classes everytime to find the Theme as well, instead of using such cache. I don't see how it would function as a cache without it being stored somewhere.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: 2 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8krmgREot54C20ll7O:bb08dxa) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

What? I'm confused. Are you talking about the map? If so, yes it IS a cache. We compute the targetRoutes on init and save the paths in the cache, so the same thing is done with the themes. We can transverse the classes everytime to find the Theme as well, instead of using such cache. I don't see how it would function as a cache without it being stored somewhere.
</blockquote></details>

There can be different types of cache:
- permanent cache
- temporal cache

At the moment the `Map` is a permanent cache.
If it's used only as a permanent cache then I vote to remove it and make a calculation on demand.
Sometimes a map is used as a temporal storage during calculation (e.g. prevent the endless loop or just gets something that is already being calculated). In this case this map normally should be passed as an argument for calculations methods (another option to use it as a field in the class to avoid extra argument but this is bad approach).

So if it's a permanent cache: I vote for calculation on demand without a cache.
If it's temporal cache ( I didn't go deeply into the calculation logic) then it should be a parameter in the methods.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: 2 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L8lVtdZ9Gufhp1JAuy9:b-l5rrzv) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
> The info could be stored in the RouteTarget.

Could you do that ?
I would like to avoid one more `AtomicReference` and map inside it as I already said since it's too easy to modify the map and it becomes not thread safe in this case.

I'm too busy with release now.
Unblocking.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 1 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r1](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L8kMSlZCOEqKM6px9mI:-L9AipMNBPPjINf3qGFk:b-s87d5z) ([raw file](https://github.com/vaadin/flow/blob/c0b4dec26fd1191409b7a364ea747cb67f14f55f/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> The info could be stored in the RouteTarget.

Could you do that ?
I would like to avoid one more `AtomicReference` and map inside it as I already said since it's too easy to modify the map and it becomes not thread safe in this case.

I'm too busy with release now.
Unblocking.
</blockquote></details>

New attempt - now there's no `AtomicReference` anymore, and the Theme is saved under the `RouteTarget` object.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 1 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 470 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9DxPQY2qWf7Ov88er2:-L9DxPQY2qWf7Ov88er3:b-w2ihle) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L470)):*
> ```Java
>             targetRoutesMap.put(navigationTarget, route);
>             routeThemesMap.put(navigationTarget,
>                     findThemeForNavigationTarget(navigationTarget));
> ```

You shouldn't collect these here as the result is not used anywhere.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 513 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9Dy-QwAigE_Em__PLv:-L9Dy-QwAigE_Em__PLw:b-o0qqtl) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L513)):*
> ```Java
> 
>         Optional<Theme> themeAnnotation = AnnotationReader
>                 .getAnnotationFor(navigationTarget, Theme.class);
> ```

Here you are still only checking the given navigation target when you should check the topParent layout for the annotation.
Navigation target is only used if it has not parent layouts.

e.g. this should be using the things `getThemeFor` has, as now it's not actually finding the theme, only checking the given target and then possibly returning default.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 574 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9E-CbABtP8DqSHp5FD:-L9E-CbABtP8DqSHp5FE:b88uroc) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L574)):*
> ```Java
>     public Optional<Class<? extends AbstractTheme>> getThemeFor(
>             Class<?> navigationTarget, String path) {
>         Class<? extends RouterLayout> topParentLayout = RouterUtil
> ```

This will never hit the predetermined route theme except when you don't have a parent layout. this should always ask the route target for the theme with the navigationTarget as the Theme should have been set from topParentLayout in initialization.

e.g. this should only do the private `getThemeFor` and findTheme should do this topParentLayout check.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 4 files at r2, 4 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 3 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 470 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9DxPQY2qWf7Ov88er2:-L9EIYaH3w4i2Jd0VsY4:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L470)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

You shouldn't collect these here as the result is not used anywhere.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 513 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9Dy-QwAigE_Em__PLv:-L9EJLTU8qb04u1Li18C:bpc4hw5) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L513)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Here you are still only checking the given navigation target when you should check the topParent layout for the annotation.
Navigation target is only used if it has not parent layouts.

e.g. this should be using the things `getThemeFor` has, as now it's not actually finding the theme, only checking the given target and then possibly returning default.
</blockquote></details>

Refactored the code, now the `path` parameter was removed and resolved internally.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 574 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9E-CbABtP8DqSHp5FD:-L9EL2lC8yjnJ_XsJqVI:bnn1lw8) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L574)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

This will never hit the predetermined route theme except when you don't have a parent layout. this should always ask the route target for the theme with the navigationTarget as the Theme should have been set from topParentLayout in initialization.

e.g. this should only do the private `getThemeFor` and findTheme should do this topParentLayout check.
</blockquote></details>

Refactored the code.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

caalador commented:
Reviewed 1 of 5 files at r5.
Review status: 4 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 470 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9DxPQY2qWf7Ov88er2:-L9Eqa8y033yNiZ8sD8s:b-464mhs) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L470)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Don't send target routes as it is actually not needed. see next down.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 574 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9E-CbABtP8DqSHp5FD:-L9Erkmr5Hs-sPCiHgq1:b-62xc04) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L574)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Refactored the code.
</blockquote></details>

Sorry. This probably should have path also.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r5](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9EqoniCfQe0G2Cl5UR:-L9EqoniCfQe0G2Cl5US:b-o8u35s) ([raw file](https://github.com/vaadin/flow/blob/a14d27a776729ca022db7abb035587171e566ddd/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
> ```Java
>             }
>             routeTarget.setThemeFor(navigationTarget,
>                     findThemeForNavigationTarget(targetRoutesMap,
> ```

instead of giving the RoutesMap you should give the alias string which is actually the path that should be used to determine the topParentLayout. 

e.g. `findThemeForNavigationTarget(alias, navigationTarget)` the aliases name is missleading as it's actually paths that contain all `RouteAlias` paths AND `Route` path.

Sorry about the trickling information stream.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 507 at r5](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9ErMKx5MEP1EDI8CgG:-L9ErMKy9ow35gFgxdZf:b-u0wrms) ([raw file](https://github.com/vaadin/flow/blob/a14d27a776729ca022db7abb035587171e566ddd/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L507)):*
> ```Java
> 
>     private Class<? extends AbstractTheme> findThemeForNavigationTarget(
>             Map<Class<? extends Component>, String> targetRoutesMap,
> ```

Should just get `String path, Class<?> navigationTarget`

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 3 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 470 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9DxPQY2qWf7Ov88er2:-L9F5UpE5y4CWiSEhyF3:bs7wxsz) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L470)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Don't send target routes as it is actually not needed. see next down.
</blockquote></details>

Done².

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 574 at r4](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9E-CbABtP8DqSHp5FD:-L9F5Zz5-WfpC4ubr220:b-flyi4m) ([raw file](https://github.com/vaadin/flow/blob/24b2d5b5c9cd331b3d573d22ea5c43c2f79f8052/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L574)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Sorry. This probably should have path also.
</blockquote></details>

Rollbacked the code.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 501 at r5](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9EqoniCfQe0G2Cl5UR:-L9F5XE08lnMb-qBUpDv:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/a14d27a776729ca022db7abb035587171e566ddd/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L501)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

instead of giving the RoutesMap you should give the alias string which is actually the path that should be used to determine the topParentLayout. 

e.g. `findThemeForNavigationTarget(alias, navigationTarget)` the aliases name is missleading as it's actually paths that contain all `RouteAlias` paths AND `Route` path.

Sorry about the trickling information stream.
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 507 at r5](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9ErMKx5MEP1EDI8CgG:-L9F5Y4v0OswrQyeJV5t:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/a14d27a776729ca022db7abb035587171e566ddd/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L507)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Should just get `String path, Class<?> navigationTarget`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

caalador commented:
Reviewed 4 of 5 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-server/src/main/java/com/vaadin/flow/server/BootstrapHandler.java, line 324 at r6](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9FDkmZC07iQgTST9bH:-L9FDkmZC07iQgTST9bI:bnxibv8) ([raw file](https://github.com/vaadin/flow/blob/25b07c64223398da3d3d4ecbdb5b7fa7beb379b3/flow-server/src/main/java/com/vaadin/flow/server/BootstrapHandler.java#L324)):*
> ```Java
>         protected Optional<Class<? extends AbstractTheme>> getTheme() {
>             if (pageConfigurationHolder == null) {
>                 return Optional.empty();
> ```

Should this return default?

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 576 at r6](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9FEfOV9yClrSwyen9l:-L9FEfOV9yClrSwyen9m:b-wu7qtg) ([raw file](https://github.com/vaadin/flow/blob/25b07c64223398da3d3d4ecbdb5b7fa7beb379b3/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L576)):*
> ```Java
>     public Optional<Class<? extends AbstractTheme>> getThemeFor(
>             Class<?> navigationTarget, String path) {
> ```

First check by path as it might target a `RouteAlias` and then check by class as it gives the `Route` route.

e.g. 
```
if(navigationTargetsInitialized()) {
  if(path != null) { 
    routeTarget = routes.get().get(path);
  }
  Map<> routes = targetRoutes.get();
  if(routeTarget == null && routes.containsKey(navigationTarget)) {
    String routePath = routes.get(navigationTarget);
    routeTarget = routes.get().get(routePath);
  }
  if(routeTarget != null) {
    return ...;
  }
}
return ...;
```

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 5 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[flow-server/src/main/java/com/vaadin/flow/server/BootstrapHandler.java, line 324 at r6](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9FDkmZC07iQgTST9bH:-L9FMzRM1cMi_zoyaXQp:b-i6qu3o) ([raw file](https://github.com/vaadin/flow/blob/25b07c64223398da3d3d4ecbdb5b7fa7beb379b3/flow-server/src/main/java/com/vaadin/flow/server/BootstrapHandler.java#L324)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Should this return default?
</blockquote></details>

The logic to find the default is buried down in the registry, so I changed the `getThemeFor` method to accept null class parameter.

---

*[flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java, line 576 at r6](https://beta.reviewable.io/reviews/vaadin/flow/3785#-L9FEfOV9yClrSwyen9l:-L9FNFhu5lHBaEGOg9NH:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/25b07c64223398da3d3d4ecbdb5b7fa7beb379b3/flow-server/src/main/java/com/vaadin/flow/server/startup/RouteRegistry.java#L576)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

First check by path as it might target a `RouteAlias` and then check by class as it gives the `Route` route.

e.g. 
```
if(navigationTargetsInitialized()) {
  if(path != null) { 
    routeTarget = routes.get().get(path);
  }
  Map<> routes = targetRoutes.get();
  if(routeTarget == null && routes.containsKey(navigationTarget)) {
    String routePath = routes.get(navigationTarget);
    routeTarget = routes.get().get(routePath);
  }
  if(routeTarget != null) {
    return ...;
  }
}
return ...;
```

</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785)*
<!-- Sent from Reviewable.io -->

caalador commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3785#-:-L9FQ6fZ2hPyvb-vx94v:bg7igk9)*
<!-- Sent from Reviewable.io -->
